pkg/da/data_availability_header.go

```diff
@@ -0,0 +1,213 @@
+	// Check that square size is with range
+	if squareSize < consts.MinSquareSize || squareSize > consts.MaxSquareSize {
+		return DataAvailabilityHeader{}, fmt.Errorf(
+			"invalid square size: min %d max %d provided %d",
+			consts.MinSquareSize,
+			consts.MaxSquareSize,
+			squareSize,
+		)
+	}
+	// check that valid number of shares have been provided
+	if squareSize*squareSize != uint64(len(shares)) {
+		return DataAvailabilityHeader{}, fmt.Errorf(
+			"must provide valid number of shares for square size: got %d wanted %d",
+			len(shares),
+			squareSize*squareSize,
+		)
+	}
+
+	tree := wrapper.NewErasuredNamespacedMerkleTree(squareSize)
+
+	// TODO(ismail): for better efficiency and a larger number shares
+	// we should switch to the rsmt2d.LeopardFF16 codec:
+	extendedDataSquare, err := rsmt2d.ComputeExtendedDataSquare(shares, rsmt2d.NewRSGF8Codec(), tree.Constructor)
+	if err != nil {
+		return DataAvailabilityHeader{}, err
+	}
+
+	// generate the row and col roots using the EDS
+	dah := DataAvailabilityHeader{
+		RowsRoots:   extendedDataSquare.RowRoots(),
+		ColumnRoots: extendedDataSquare.ColRoots(),
+	}
+
+	// generate the hash of the data using the new roots
+	dah.Hash()
+
+	return dah, nil
+}
+
+// String returns hex representation of merkle hash of the DAHeader.
+func (dah *DataAvailabilityHeader) String() string {
+	if dah == nil {
+		return "<nil DAHeader>"
+	}
+	return fmt.Sprintf("%X", dah.Hash())
+}
+
+// Equals checks equality of two DAHeaders.
+func (dah *DataAvailabilityHeader) Equals(to *DataAvailabilityHeader) bool {
+	return bytes.Equal(dah.Hash(), to.Hash())
+}
+
+// Hash computes and caches the merkle root of the row and column roots.
+func (dah *DataAvailabilityHeader) Hash() []byte {
+	if dah == nil {
+		return merkle.HashFromByteSlices(nil)
+	}
+	if len(dah.hash) != 0 {
+		return dah.hash
+	}
+
+	colsCount := len(dah.ColumnRoots)
+	rowsCount := len(dah.RowsRoots)
+	slices := make([][]byte, colsCount+rowsCount)
+	for i, rowRoot := range dah.RowsRoots {
+		slices[i] = rowRoot
+	}
+	for i, colRoot := range dah.ColumnRoots {
+		slices[i+colsCount] = colRoot
+	}
+	// The single data root is computed using a simple binary merkle tree.
+	// Effectively being root(rowRoots || columnRoots):
+	dah.hash = merkle.HashFromByteSlices(slices)
+	return dah.hash
+}
+
+func (dah *DataAvailabilityHeader) ToProto() (*daproto.DataAvailabilityHeader, error) {
+	if dah == nil {
+		return nil, errors.New("nil DataAvailabilityHeader")
+	}
+
+	dahp := new(daproto.DataAvailabilityHeader)
+	dahp.RowRoots = dah.RowsRoots
+	dahp.ColumnRoots = dah.ColumnRoots
+	return dahp, nil
+}
+
+func DataAvailabilityHeaderFromProto(dahp *daproto.DataAvailabilityHeader) (dah *DataAvailabilityHeader, err error) {
+	if dahp == nil {
+		return nil, errors.New("nil DataAvailabilityHeader")
+	}
+
+	dah = new(DataAvailabilityHeader)
+	dah.RowsRoots = dahp.RowRoots
+	dah.ColumnRoots = dahp.ColumnRoots
+
+	return dah, dah.ValidateBasic()
```

Comment:
related to #282, we will now throw an error if the square size is not within range while passing a DAH over the wire